### PR TITLE
blogpost: Multi-version upgrades

### PR DESCRIPTION
**Pull request by efritz**
This PR adds a technical deep-dive into how we developed multi-version upgrades.

### Test

1. Ensure prettier has standardized the proposed changes.

### CODE REVIEW

**Comment by efritz** [comment]
@iskyOS Can we get the non-screenshot images here done in a better style? Once we have those I can remove them from git here and add them to the public GCS assets bucket.

**Comment by danieldides** [comment]
nit: Typically in commit/version graphs like this I see the arrows pointing the opposite direction, where each commit would reference its parent only since it's unaware of its decedents. Not sure this is worth fixing or matters. 🤷‍♀️

**Comment by efritz** [comment]
@danieldides Yeah I agree. @iskyOS Can we ensure new art reverses the direction of the arrows in the graphs? If they point right to left it would be a more intuitive as each migration points to its parent.

**Review by vdavid — COMMENTED** [review]
I only skimmed through it, but it was an awesome geeky read, both informative and enjoyable.
Meta (no action needed, just wanted to complain to the void): This is my first about page review, and I wish I had a nice rendered preview of the article.

**Comment by efritz** [comment]
@vdavid Preview can be found in the [deploy/netlify](https://deploy-preview-5723--sourcegraph.netlify.app/blog/multi-version-upgrades) action on this PR.

File: content/blogposts/2022/multi-version-upgrades.md
Comment (efritz) [reply content/blogposts/2022/multi-version-upgrades.md:39]: The sentence is correct but really clunky. Does `Note that we are unable to squash migration definitions that are undefined in the source version of an upgrade.` read better?

Comment (vdavid) [reply content/blogposts/2022/multi-version-upgrades.md:39]: Oh, I see now what you meant there. Then the original phrasing is more precise, but I find your suggested alternative easier to understand.

Comment (efritz) [reply content/blogposts/2022/multi-version-upgrades.md:39]:
```suggestion
Note that we are unable to squash migration definitions that are undefined in the source version of an upgrade. Doing so would make it impossible for the previous release to determine _which_ migration definitions to apply. Squashing in the manner described above guarantees that none of the migration definitions in the upgrade path from the previous minor version would be altered.
```

**Comment by vdavid** [comment]
Ah, never mind, I understand it just comes up later in the queue. Thanks, TIL.

**Comment by efritz** [comment]
@vdavid Netlify is currently processing from the last commit.

image

File: content/blogposts/2022/multi-version-upgrades.md
Comment (vdavid) [inline content/blogposts/2022/multi-version-upgrades.md:3]: Bringing attention to this to make sure we don't forget updating this as well as the `heroImage` and `socialImage` before we merge this.

File: content/blogposts/2022/multi-version-upgrades.md
Comment (efritz) [inline content/blogposts/2022/multi-version-upgrades.md:21]: Ensure we update the placeholder link targett.

**Review by chrismwendt — DISMISSED** [review]

File: content/blogposts/2022/multi-version-upgrades.md
Comment (iskyOS) [reply content/blogposts/2022/multi-version-upgrades.md:21]: https://about.sourcegraph.com/blog/release/4.0

File: content/blogposts/2022/multi-version-upgrades.md
Comment (iskyOS) [inline content/blogposts/2022/multi-version-upgrades.md:79]:
```suggestion
We deal with the **Wrench of Concurrent Indexes** in a similar way. Migration definitions squashed away before the introduction of our special-casing of concurrent index creation will miss the markers in their metadata. In this case we only have to spot rewrite migrations for the `frontend` schema, but there were a significant number of them ([1528395696](https://github.com/sourcegraph/sourcegraph/pull/12591), [1528395707](https://github.com/sourcegraph/sourcegraph/pull/13217), [1528395708](https://github.com/sourcegraph/sourcegraph/pull/13217), [1528395736](https://github.com/sourcegraph/sourcegraph/pull/14848), [1528395797](https://github.com/sourcegraph/sourcegraph/pull/19053), [1528395877](https://github.com/sourcegraph/sourcegraph/pull/24798), [1528395878](https://github.com/sourcegraph/sourcegraph/pull/24798), [1528395886](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395887](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395888](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395893](https://github.com/sourcegraph/sourcegraph/pull/25366), [1528395894](https://github.com/sourcegraph/sourcegraph/pull/25360), [1528395896](https://github.com/sourcegraph/sourcegraph/pull/25405), [1528395897](https://github.com/sourcegraph/sourcegraph/pull/25490), [1528395899](https://github.com/sourcegraph/sourcegraph/pull/25520), [1528395900](https://github.com/sourcegraph/sourcegraph/pull/25520), [1528395935](https://github.com/sourcegraph/sourcegraph/pull/27359), [1528395936](https://github.com/sourcegraph/sourcegraph/pull/27359), and [1528395954](https://github.com/sourcegraph/sourcegraph/pull/28387)).
```

Comment (iskyOS) [reply content/blogposts/2022/multi-version-upgrades.md:79]: nubmer -> number

File: content/blogposts/2022/multi-version-upgrades.md
Comment (iskyOS) [inline content/blogposts/2022/multi-version-upgrades.md:7]: Just adding a note to update publish date when we're ready

**Review by iskyOS — COMMENTED** [review]
Looks great! Added a couple small comments (one typo), but no changes recommended

**Comment by efritz** [comment]
@iskyOS Thanks! I'll make these changes once I update the linked images as well.

**Review by DaedalusG — APPROVED** [review]
Love the charts 🔥

File: content/blogposts/2022/multi-version-upgrades.md
Comment (efritz) [reply content/blogposts/2022/multi-version-upgrades.md:3]: Updated both to https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/mvu-hero.png.

File: content/blogposts/2022/multi-version-upgrades.md
Comment (jdorfman) [inline content/blogposts/2022/multi-version-upgrades.md:69]: This isn't loading, needs to be uploaded to gcloud @efritz

Comment (iskyOS) [reply content/blogposts/2022/multi-version-upgrades.md:69]: We also have new guidance on embedding images using <Figure>: https://about.sourcegraph.com/blog/starter-pack#video

Comment (iskyOS) [reply content/blogposts/2022/multi-version-upgrades.md:69]:
```suggestion
<Figure
    src="blog/image name"
    alt=""
    caption=""
  />
```

Comment (efritz) [reply content/blogposts/2022/multi-version-upgrades.md]: Updated these.

Comment (efritz) [reply content/blogposts/2022/multi-version-upgrades.md]: @jdorfman I think you were looking at an old version.

File: content/blogposts/2022/multi-version-upgrades.md
Comment (iskyOS) [inline content/blogposts/2022/multi-version-upgrades.md:35]:
```suggestion
Because this problem would only get worse with time, we began to squash migration definitions by replacing the set of definitions at the head of the sequence with a single (yet equivalent) migration file. For Sourcegraph v3.9.0, we [squashed definitions](https://github.com/sourcegraph/sourcegraph/pull/5967) that were also defined in Sourcegraph v3.7.0 into a single file. This cut startup time with an empty database from 20s to 5s and also reduced the chance of hitting the startup timeout in our testing environment.
```

File: content/blogposts/2022/multi-version-upgrades.md
Comment (efritz) [inline content/blogposts/2022/multi-version-upgrades.md:7]:
```suggestion
publishDate: 2022-10-20T10:00-07:00
```

@iskyOS Will leave this to you to choose a publish date :)

File: content/blogposts/2022/multi-version-upgrades.md
Comment (iskyOS) [inline content/blogposts/2022/multi-version-upgrades.md:57]:
```suggestion
Well that seemed like a bunch of work for no payoff. **Why do that at all?** Short answer: engineering ergonomics. When we had a gapless sequence we would **frequently** run into merge conflicts when two engineers add migrations around the same time. Since each branch did not contain the other's migration, they would both choose the same sequence number. Because of limitations in our CI (no build queues + no non-stale base branch merge requirement), this conflict would, the majority of the time, be caught only after both branches have been merged and validation errors turned the `main` branch to go red.
```

File: content/blogposts/2022/multi-version-upgrades.md
Comment (iskyOS) [inline content/blogposts/2022/multi-version-upgrades.md:69]: @efritz What image should this be?

Comment (efritz) [reply content/blogposts/2022/multi-version-upgrades.md:69]: https://github.com/sourcegraph/about/blob/8eda68889b3f6db9cb4638003ee6ca64e4a7c7b5/public/blog/temp/timeline.png

Comment (efritz) [reply content/blogposts/2022/multi-version-upgrades.md:69]:
```suggestion
<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/timeline.png" alt="timeline" />
```

File: content/blogposts/2022/multi-version-upgrades.md
Comment (iskyOS) [reply content/blogposts/2022/multi-version-upgrades.md:107]:
```suggestion
We've continued to iterate on our schema and data migration infrastructure to remove entire classes of pain from our customers, and we are taking significant strides to decrease maintenance burden. Keep your eyes peeled for future improvements in this area.
```

**Review by jdorfman — APPROVED** [review]
LOVE IT

**Pull request merged**
